ir/src/integrity_stmts/constraint.rs

```diff
@@ -0,0 +1,77 @@
+/// trace in both the current and the next rows then the row offset will be 1.
+#[derive(Debug, Clone)]
+pub struct ConstraintRoot {
+    index: NodeIndex,
+    domain: ConstraintDomain,
+}
+
+impl ConstraintRoot {
+    /// Creates a new [ConstraintRoot] with the specified entry index and row offset.
+    pub fn new(index: NodeIndex, domain: ConstraintDomain) -> Self {
+        Self { index, domain }
+    }
+
+    /// Returns the index of the entry node of the subgraph representing the constraint.
+    pub fn node_index(&self) -> &NodeIndex {
+        &self.index
+    }
+
+    /// Returns the [ConstraintDomain] for this constraint, which specifies the rows against which
+    /// the constraint should be applied.
+    pub fn domain(&self) -> ConstraintDomain {
+        self.domain
+    }
+}
+
+/// The domain to which the constraint is applied, which is either the first or last row (for
+/// boundary constraints), every row (for validity constraints), or every frame (for transition
+/// constraints). When the constraint is applied to a frame the inner value specifies the size of
+/// the frame. For example, for a transition constraint that is applied against the current and next
+/// rows, the frame size will be 2.
+#[derive(Debug, Clone, PartialEq, Eq, Copy)]
+pub enum ConstraintDomain {
+    FirstRow,          // for boundary constraints against the first row
+    LastRow,           // for boundary constraints against the last row
+    EveryRow,          // for validity constraints
+    EveryFrame(usize), // for transition constraints
+}
+
+impl ConstraintDomain {
+    /// Combines the two [ConstraintDomain]s into a single [ConstraintDomain] that represents the
+    /// maximum constraint domain. For example, if one domain is [ConstraintDomain::EveryFrame(2)]
+    /// and the other is [ConstraintDomain::EveryFrame(3)] then the result will be
+    /// [ConstraintDomain::EveryFrame(3)].
+    pub fn merge(&self, other: &ConstraintDomain) -> ConstraintDomain {
+        if self == other {
+            return *other;
+        }
+
+        match (self, other) {
+            (ConstraintDomain::EveryFrame(a), ConstraintDomain::EveryFrame(b)) => {
+                ConstraintDomain::EveryFrame(*a.max(b))
+            }
+            (ConstraintDomain::EveryFrame(a), _) => ConstraintDomain::EveryFrame(*a),
+            (_, ConstraintDomain::EveryFrame(b)) => ConstraintDomain::EveryFrame(*b),
+            // for any other pair of constraints which are not equal, the result of combining the
+            // domains is to apply the constraint at every row.
+            _ => ConstraintDomain::EveryRow,
+        }
+    }
+}
+
+impl From<usize> for ConstraintDomain {
+    /// Creates a [ConstraintDomain] from the specified row offset.
+    fn from(row_offset: usize) -> Self {
+        if row_offset == 0 {
+            ConstraintDomain::EveryRow
+        } else {
+            ConstraintDomain::EveryFrame(row_offset + 1)
```

Comment:
question: What is the reasoning behind doing "row_offset + 1" here? I couldn't figure it out.

Reply:
The size of the frame is the number of rows in the frame, so we need to add one to count the "current row" with index zero. For example, when row_offset is 1 our frame size is 2

Reply:
Ah, got it, thanks.